Core/Src/main.c

```diff
@@ -1,181 +1,224 @@
-/**
-  * @brief  Reports the name of the source file and the source line number
-  *         where the assert_param error has occurred.
-  * @param  file: pointer to the source file name
-  * @param  line: assert_param error line source number
-  * @retval None
-  */
-void assert_failed(uint8_t *file, uint32_t line)
-{
-  /* USER CODE BEGIN 6 */
-  /* User can add his own implementation to report the file name and line number,
-     ex: printf("Wrong parameters value: file %s on line %d\r\n", file, line) */
-  /* USER CODE END 6 */
-}
-#endif /* USE_FULL_ASSERT */
-
-/************************ (C) COPYRIGHT STMicroelectronics *****END OF FILE****/
+/* USER CODE BEGIN Header */
+/**
+  ******************************************************************************
+  * @file           : main.c
+  * @brief          : Main program body
+  ******************************************************************************
+  * @attention
+  *
+  * <h2><center>&copy; Copyright (c) 2021 STMicroelectronics.
+  * All rights reserved.</center></h2>
+  *
+  * This software component is licensed by ST under BSD 3-Clause license,
+  * the "License"; You may not use this file except in compliance with the
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+#include <stdio.h>
+#include <string.h>
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+
+#define MAX_ADC_VALUE 1024
+#define MIN_DUTY_CYCLE_COUNT 100 /* 5% of timer period */
+#define MAX_DUTY_CYCLE_COUNT 200 /* 10% of timer period */
+
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+  uint8_t start_byte = 1;
+  uint8_t spi_recv_buf[2] = {0};

```

Comment:
Close but you are missing some details here. Please take another look at the image below. You can see the part I circled, that there are a total of 3 bytes being sent and received at the same time (not only 1 byte sent and 2 bytes received). 

For the 3 bytes you must send, you have to send the start byte (which you already have there) you need to send the configuration byte (that contains bits D2, D1, and D0, and then you need to send a "don't care" byte which means you can send anything you want, it doesn't matter (so maybe all 0s).

For the bytes you receive, you will first receive a byte of random (which means garbage) bits then you will receive a byte that has garbage bits but then B9 and B8, then you will receive a byte that contains the rest of B7-B0. So if you want to get the whole adc value in 1 uint16_t variable, you will have to do some bit manipulation using bitwise operators (which it looks like you already do down below!)

So in summary it looks like you will need a 3 element buffer for both send and receive right?

![image](https://user-images.githubusercontent.com/69175242/193465776-ad088f1e-b6d5-404d-a3e0-b4e15dc72bef.png)